[1.21.11] Allow custom `GameRuleType`s to be defined

ApexModder opened the pull request:
This PR marks `GameRuleType` as a extensible enum allowing custom rule types to be defined.

Game rule tests in `LevelTests` have been expanded to register and modify 2 new game rules. `custom_double_game_rule` and `custom_string_game_rule`. Very basic translations have also been provided for these as the edit game rule screen was starting to be a pain with all the raw ids

I have also tried playing around with adding a new `GameRuleCategory` which _just worked_ out of the box, only thing of note we might want to add a new event for registering them to the backing ArrayList.

While custom these custom types work just fine i have noticed that they do not currently appear in the game rules screen when creating a new world.
After some digging this is due to that screen only listening to the `visitBoolean` and `visitInteger` visitor methods when adding entries to the game rule widget list. On top of that custom "edit box" widgets may/may not be required depending on the game rule type, will look more into this later though.

ApexModder committed:
Allow custom `GameRuleType`s to be defined

ApexModder committed:
All registering custom game rule edit entries

Add default generic game rule entry

ApexModder commented:
PR has been updated to allow for the registration of custom `EditGameRulesScreen.RuleEntry` factories, as well as providing a generic `RuleEntry` class which is used when no factory has been registered.

With this change the new `double` and `string` game rules now display in the `EditGameRulesScreen`

<img width="1729" height="1039" alt="image" src="https://github.com/user-attachments/assets/6dfcee42-c6e2-4612-8b5f-d43f9bd324bc" />

XFactHD commented on src/client/java/net/neoforged/neoforge/client/gui/widget/gamerule/GameRuleEntry.java:
```diff
@@ -0,0 +1,53 @@
+        return children;
+    }
+
+    protected void renderLabel(GuiGraphics graphics, int mouseX, int mouseY) {
```
```suggestion
    protected void renderLabel(GuiGraphics graphics, int y, int x) {
```

ApexModder replied:
huh didnt notice that the names of these are flipped from most other x,y pairings. Mojang why is it y,x ?!?

XFactHD commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1120,4 +1129,17 @@ private static boolean isValidDebugEntryForSearch(String searchText, Identifier
+        }
+
+        return new GenericGameRuleEntry<>(screen, label, tooltip, str, gameRule);
+    }
```
I would prefer moving this to a `GameRuleEntryFactoryManager`, having all these random fields sprayed all over this class is a nightmare.

XFactHD commented on src/client/java/net/neoforged/neoforge/client/event/RegisterGameRuleEntryFactoryEvent.java:
```diff
@@ -0,0 +1,56 @@
+    @FunctionalInterface
+    public interface Factory<T> {
+        EditGameRulesScreen.RuleEntry create(EditGameRulesScreen screen, Component description, List<FormattedCharSequence> tooltip, String str, GameRule<T> gameRule);
+    }
```
Nitpick: move this to a top-level class next to `GameRuleEntry`

ApexModder replied:
Moved into its own class `GameRuleEntryFactory` under the new `.client.gamerules` package

XFactHD commented on patches/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java.patch:
```diff
@@ -0,0 +1,15 @@
+                     }
+ 
++                    @Override
++                    public <T> void visit(GameRule<T> p_461081_) {
```
As far as I can tell, this is also called for bool and int rules, so those would need to be filtered out.

ApexModder replied:
This is true, the generic `visit(GameRule<T>)` is called for all game rules and the type specific vistors are triggered via the `GameRule`s visitor callback, which custom game rule types should be leaving as a empty callback unless your specific type needs some special visitor.

I even made note of this in the [`LevelTests`](https://github.com/ApexModder/NeoForge/blob/a1cdbc90c8294c5418e1838c9ea31a30358975c9/tests/src/main/java/net/neoforged/neoforge/debug/level/LevelTests.java#L66-L68) class, odd that i forgot about filtering out the vanilla types here

ApexModder committed:
Move everything to a `gamerules` package

Move game rule entry factory into its own class
Rename fields to better match their usage

ApexModder committed:
Flip parameter names in GameRuleEntry

ApexModder committed:
Filter out bool/int rule types when appending custom entries

ApexModder committed:
Introduce event for registering custom gamerule categories

ApexModder committed:
Add missing deprecated annotation

XFactHD commented on src/client/java/net/neoforged/neoforge/client/gamerules/RegisterGameRuleEntryFactoryEvent.java:
```diff
@@ -0,0 +1,46 @@
+ * This event is fired on the mod-specific event bus, only on the {@link LogicalSide#CLIENT logical client}.
+ * <p>
+ * While you are not required to register a custom factory for your {@link GameRuleType}, it is recommended to do so
+ * of your game rule requires more than a simple {@link EditBox}. For example a togglable types similar to {@link GameRuleType#BOOL Boolean}
```
```suggestion
 * if your game rule requires more than a simple {@link EditBox}. For example a togglable type similar to {@link GameRuleType#BOOL Boolean}
```

ApexModder committed:
Update src/client/java/net/neoforged/neoforge/client/gamerules/RegisterGameRuleEntryFactoryEvent.java

Co-authored-by: Dennis C <xfacthd@gmx.de>

XFactHD reviewed (commented):
Overall looks good to me, though we need to make sure the addition of custom rule types works properly with the management API.

ApexModder commented:
After some googling of figuring out some very basic JsonRPC things, I was able to update the value of `neotests_custom_game_rule:custom_string_game_rule` just fine which is not only a custom game rule but also a custom rule type `string`.

So I think it should be safe to assume that the ManagementAPI works just fine with custom rule types. That being said if @Minecraftschurli or someone else who knows the JsonRPC and ManagementAPI better wants to give it better test that would be appreciated

[Screencast_20251219_152500.webm](https://github.com/user-attachments/assets/63669532-d9d2-4ae9-af9f-bdbcc97b68b3)
